scipy/stats/_distn_infrastructure.py

```diff
@@ -3836,3 +3884,423 @@ def get_distribution_names(namespace_pairs, rv_base_class):
+                   f"and the domain of the distribution is not finite. Please "
+                   f"provide finite bounds for shape `{name}` in "
+                   f"`{bounds_type}`.")
+        raise ValueError(message)
+
+    return bounds
+
+
+class FitResult:
+    r"""Result of fitting a discrete or continuous distribution to data
+
+    Attributes
+    ----------
+    dist : scipy.stats.rv_continuous or scipy.stats.rv_discrete
+        The distribution of which parameters were fit.
+    data : 1D array_like
+        The data to which the distribution was fit.
+    nllf : callable
+        The negative log-likehood function for the given `data`. Accepts a
+        single tuple containing the shapes, location, and scale of the
+        distribution.
+    params : namedtuple
+        A namedtuple containing the maximum likelihood estimates of the
+        shape parameters, location, and (if applicable) scale of the
+        distribution.
+    objective_val : float
+        The value of the negative log likelihood function evaluated with
+        the fitted shapes, i.e. ``result.nllf(result.params)``.
+    success : bool or None
+        Whether the optimizer considered the optimization to terminate
+        successfully or not.
+    message : str or None
+        Any status message provided by the optimizer.
+
+    """
+
+    def __init__(self, dist, data, discrete, nllf, res):
+        self.dist = dist
+        self.data = data
+        self.discrete = discrete
+        self.pxf = getattr(dist, "pmf", None) or getattr(dist, "pdf", None)
+        self.success = getattr(res, "success", None)
+        self.message = getattr(res, "message", None)
+
+        shape_names = [] if dist.shapes is None else dist.shapes.split(", ")
+        if getattr(dist, "pdf", False):
+            FitShapes = namedtuple('FitShapes', shape_names + ['loc', 'scale'])
+        elif getattr(dist, "pmf", False):
+            FitShapes = namedtuple('FitShapes', shape_names + ['loc'])
+
+        self.params = FitShapes(*res.x)
+        self.objective_val = getattr(res, "fun", None)
+        if self.objective_val is None:
+            self.objective_val = nllf(res.x)
+        self.nllf = nllf
+
+
+    def __repr__(self):
+        keys = ["dist", "data", "nllf", "params", "objective_val",
+                "success", "message"]
+        m = max(map(len, keys)) + 1
+        return '\n'.join([key.rjust(m) + ': ' + repr(getattr(self, key))
+                          for key in keys if getattr(self, key) is not None])
+
+    def plot(self):
+        """Visualize the fit result.
+
+        Superposes the PDF/PMF of the fitted distribution over a normalized
+        histogram of the data.
+
+        Available only if ``matplotlib`` is installed.
+
+        Returns
+        -------
+        fig : matplotlib.figure.Figure
+            A matplotlib Figure object.
+        ax : matplotlib.axes.Axes
+            A matplotlib Axes object.
```

Comment:
I haven't looked at how other SciPy functions do this sort of thing. Also, I'm not an expert `matplotlib` user, so someone should review this for best practices. 

To keep the code simple, my hope is that this function will have _no inputs_ but that the user will be able to customize everything as desired using the returned objects.

Reply:
We can have a look at this point later, long story short: usually you would want to allow the user to pass `ax` to allow easier composition (see https://github.com/statsmodels/statsmodels/blob/55f9e73b7665bc6bda0f4c13a1ac27d60c021777/statsmodels/graphics/functional.py#L298 for instance) and yes it should return `ax` as you did. But I think it should not call `plt.show` and leave this to the user. It's common to call mpl code and when you are done call yourself `plt.show`.

Reply:
Ok. Is it enough to return just the `ax` or should it return the `fig`, too? 
Do you know how to make this show up in the rendered docs?